### DISCUSSION TITLE
Use While activity in a Flowchart

### DISCUSSION THREAD

**Discussion by rosca-sabina** [elsa-workflows/elsa-core#5691] [Q&A] [answered]
Hi everyone,

I have a workflow with a Flowchart as the root activity, and I'm having some trouble integrating a While activity into the workflow.

The While activity has a custom activity as its Body, and this custom activity can complete with different outcomes. I'd like that when the exit condition for the While is met, the While to complete with the same outcome that that the body activity completed with, so that I can configure the Flowchart to go from the While to another activity based on that outcome.

I tried setting a custom behavior on the While, and wrapping the While in another custom activity, but I can't quite get there.

I've provided a small example below. In the example, after counter reaches 3 I want the workflow to move on to `writeLineActivityA`, because `TestActivity` set the outcome as `"A"`.

Thanks in advance for the help!
``` csharp
public class TestWorkflow : WorkflowBase
{
    protected override void Build(IWorkflowBuilder builder)
    {
        var counterVariable = builder.WithVariable<int>(0);

        var whileActivity = new While(() => true)
        {
            Body = new Sequence
            {
                Activities =
                [
                    new TestActivity(counterVariable),
                    new If
                    {
                        Condition = new(context => counterVariable.Get(context)! >= 3),
                        Then = new Break(),
                        Else = new WaitActivity(TimeSpan.FromSeconds(5))
                    }
                ]
            }
        };

        builder.Root = new Flowchart
        {
            CanStartWorkflow = true,
            Start = whileActivity,
            Activities =
            [
                whileActivity,
                writeLineActivityA,
                writeLineActivityB
            ],
            Connections =
            [
                new Connection(new Endpoint(whileActivity, "A"), new Endpoint(writeLineActivityA)),
                new Connection(new Endpoint(whileActivity, "B"), new Endpoint(writeLineActivityB))
            ]
        };
    }
}

public class TestActivity(Variable<int> counterVariable) : Activity<string>
{
    public Input<int> Counter { get; init; } = new(counterVariable);

    protected override async ValueTask ExecuteAsync(ActivityExecutionContext context)
    {
        var counterValue = Counter.Get(context);
        counterValue++;

        counterVariable.Set(context, counterValue);

        if (counterValue >= 3)
            await context.CompleteActivityWithOutcomesAsync("A");
        else
            await context.CompleteActivityAsync();
    }
}
```

**Comment by glime-ai** [comment, marked as answer]
To achieve the desired behavior where the `While` activity completes with the same outcome as the `TestActivity`, you need to capture the outcome of the `TestActivity` and use it to conditionally break out of the `While` loop with that outcome. 

In your current implementation, you are using a `Break` activity within an `If` condition to exit the loop when the counter reaches 3, but you are not capturing the outcome of the `TestActivity`. 

Here's an updated version of your workflow that captures the outcome of `TestActivity` and uses it to exit the `While` loop with the same outcome:

```csharp
public class TestWorkflow : WorkflowBase
{
    protected override void Build(IWorkflowBuilder builder)
    {
        var counterVariable = builder.WithVariable<int>("Counter", 0);
        var outcomeVariable = builder.WithVariable<string>("Outcome");

        var testActivity = new TestActivity
        {
            Counter = new(counterVariable)
        };

        var whileActivity = new While(context => true)
        {
            Body = new Sequence
            {
                Activities =
                [
                    testActivity,
                    new If
                    {
                        Condition = new(context => counterVariable.Get(context) >= 3),
                        Then = new Outcome("A"),
                        Else = new WaitActivity(TimeSpan.FromSeconds(5))
                    }
                ]
            }
        };

        builder.Root = new Flowchart
        {
            CanStartWorkflow = true,
            Start = whileActivity,
            Activities =
            [
                whileActivity,
                writeLineActivityA,
                writeLineActivityB
            ],
            Connections =
            [
                new Connection(new Endpoint(whileActivity, "A"), new Endpoint(writeLineActivityA)),
                new Connection(new Endpoint(whileActivity, "B"), new Endpoint(writeLineActivityB))
            ]
        };
    }
}

public class TestActivity : Activity
{
    public Input<int> Counter { get; init; }

    protected override async ValueTask<IActivityExecutionResult> OnExecuteAsync(ActivityExecutionContext context)
    {
        var counterValue = Counter.Get(context);
        counterValue++;

        context.SetVariable("Counter", counterValue);

        if (counterValue >= 3)
        {
            context.SetVariable("Outcome", "A");
            return Outcome("A");
        }
        else
        {
            return Outcome("Continue");
        }
    }
}
```

In this updated version, I've added an `outcomeVariable` to capture the outcome of `TestActivity`. The `TestActivity` now sets the outcome in the context, and the `While` activity's `If` condition uses an `Outcome` activity to exit the loop with the outcome "A". The `Flowchart` connections are then based on the outcome of the `While` activity.

Please note that the `Outcome` activity is a hypothetical activity that you would need to implement to set the outcome of the `While` loop based on the `TestActivity`'s result. Elsa does not have an `Outcome` activity out of the box, so you would need to create a custom activity that sets the outcome of the parent activity (`While` in this case) based on a condition or a variable.

Remember to adjust the code to fit the actual API of the Elsa Workflow library you are using, as the code provided is a conceptual example and may not directly compile without the appropriate custom activity implementations.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=dRbvTb7IhCSYiI&utm_medium=github).

**Comment by rosca-sabina** [comment]
Thanks! Could you show me how to implement the Outcome activity, which sets the outcome of the parent activity based on the value of a variable?
